lib/bolt/shell/powershell.rb

```diff
@@ -0,0 +1,263 @@
+        @extensions = DEFAULT_EXTENSIONS + extensions
+      end
+
+      def provided_features
+        ['powershell']
+      end
+
+      def default_input_method(executable)
+        powershell_file?(executable) ? 'powershell' : 'both'
+      end
+
+      def powershell_file?(path)
+        File.extname(path).downcase == '.ps1'
+      end
+
+      def validate_extensions(ext)
+        unless @extensions.include?(ext)
+          raise Bolt::Node::FileError.new("File extension #{ext} is not enabled, "\
+                                          "to run it please add to 'winrm: extensions'", 'FILETYPE_ERROR')
+        end
+      end
+
+      def process_from_extension(path)
+        case Pathname(path).extname.downcase
+        when '.rb'
+          [
+            'ruby.exe',
+            %W[-S "#{path}"]
+          ]
+        when '.ps1'
+          [
+            'powershell.exe',
+            [*PS_ARGS, path]
+          ]
+        when '.pp'
+          [
+            'puppet.bat',
+            %W[apply "#{path}"]
+          ]
+        else
+          # Run the script via cmd, letting Windows extension handling determine how
+          [
+            'cmd.exe',
+            %W[/c "#{path}"]
+          ]
+        end
+      end
+
+      def escape_arguments(arguments)
+        arguments.map do |arg|
+          if arg =~ / /
+            "\"#{arg}\""
+          else
+            arg
+          end
+        end
+      end
+
+      def set_env(arg, val)
+        "[Environment]::SetEnvironmentVariable('#{arg}', @'\n#{val}\n'@)"
+      end
+
+      def quote_string(string)
+        "'" + string.gsub("'", "''") + "'"
+      end
+
+      def write_executable(dir, file, filename = nil)
+        filename ||= File.basename(file)
+        validate_extensions(File.extname(filename))
+        remote_path = "#{dir}\\#{filename}"
+        conn.copy_file(file, remote_path)
+        remote_path
+      end
+
+      def execute_process(path, arguments, stdin = nil)
+        quoted_args = arguments.map { |arg| quote_string(arg) }.join(' ')
+
+        quoted_path = if path =~ /^'.*'$/ || path =~ /^".*"$/
+                        path
+                      else
+                        quote_string(path)
+                      end
+        exec_cmd =
+          if stdin.nil?
+            "& #{quoted_path} #{quoted_args}"
+          else
+            <<~STR
+            $command_stdin = @'
+            #{stdin}
+            '@
+
+            $command_stdin | & #{quoted_path} #{quoted_args}
+            STR
+          end
+        Snippets.execute_process(exec_cmd)
+      end
+
+      def mkdirs(dirs)
+        mkdir_command = "mkdir -Force #{dirs.uniq.sort.join(',')}"
+        result = execute(mkdir_command)
+        if result.exit_code != 0
+          message = "Could not create directories: #{result.stderr.string}"
+          raise Bolt::Node::FileError.new(message, 'MKDIR_ERROR')
+        end
+      end
+
+      def make_tempdir
```

Comment:
Does it make sense to move the tempdir methods to a dedicated tmpdir class like the Bash shell has?

Reply:
`Bash::Tmpdir` primarily exists so we have somewhere to keep track of who currently owns the directory, in support of `run-as`. I don't think powershell has a similar need.